openprescribing/media/js/src/measure_utils.js

```diff
@@ -499,6 +500,79 @@ var utils = {
     return chOptions;
   },
 
+  startDataDownload: function(allChartData, chartId) {
+    var chartData = this.getChartDataById(allChartData, chartId);
+    var dataTable = this.getChartDataAsTable(chartData);
+    var csvData = this.formatTableAsCSV(dataTable);
+    var filename = this.sanitizeFilename(chartData.chartTitle) + '.csv';
+    downloadjs(csvData, filename, 'text/csv');
+  },
+
+  getChartDataById: function(allChartData, chartId) {
+    for (var i = 0; i<allChartData.length; i++) {
+      if (allChartData[i].chartId === chartId) {
+        return allChartData[i];
+      }
+    }
+    throw 'No matching chartId: ' + chartId;
+  },
+
+  getChartDataAsTable: function(chartData) {
+    var headers = ['date', 'numerator', 'denominator', 'ratio', 'percentile', 'cost_savings'];
+    var keyPercentiles = [10, 20, 30, 40, 50, 60, 70, 80, 90];
+    headers = headers.concat(keyPercentiles.map(function(n) { return n + 'th percentile'; }));
+    var percentilesByDate = this.groupPercentilesByDate(chartData.globalCentiles, keyPercentiles);
+    var table = chartData.data.map(function(d) {
+      return [
+          d.date, d.numerator, d.denominator, d.calc_value, d.percentile, d.cost_savings
+        ]
+        .concat(percentilesByDate[d.date]);
+    });
+    table.unshift(headers);
+    return table;
+  },
+
+  groupPercentilesByDate: function(globalCentiles, keyPercentiles) {
+    var percentilesByDate = {};
+    _.each(keyPercentiles, function(percentile) {
+      _.each(globalCentiles[percentile], function(percentileData) {
+        var date = percentileData.date;
+        if ( ! percentilesByDate[date]) {
+          percentilesByDate[date] = [];
+        }
+        percentilesByDate[date].push(percentileData.y);
+      });
+    });
+    return percentilesByDate;
+  },
+
+  formatTableAsCSV: function(table) {
+    return table.map(this.formatRowAsCSV.bind(this)).join('\n');
+  },
+
+  formatRowAsCSV: function(row) {
+    return row.map(this.formatCellAsCSV.bind(this)).join(',');
+  },
+
+  formatCellAsCSV: function(cell) {
+    cell = cell ? cell.toString() : '';
+    if (cell.match(/[,"\r\n]/)) {
```

Comment:
Is this actually required? I guess it's useful future proofing even if not.

Reply:
No, it's not required at the moment. The most likely use case would be if we start putting something more descriptive in the column headings. But in any case, my OCD won't let me write a function called "format as CSV" if it only formats as CSV by accident :)